rfc/rfc-56/rfc-56.md

```diff
@@ -0,0 +1,231 @@
+
+Here we need to pay attention to two details:
+
+Firstly, There may be a concurrency issue between reading and writing the same MARKER file causing EOFException. In 
+order to simplify the model, we think that the result is empty when catch any exception during reading specific MARKER 
+and let the subsequent checker to make up for it. For example, Writer1 will read writer2's MARKER0 file, but at this 
+moment writer2 is overwriting his MARKER0 file. Finally we will get empty set after writer1 reading writer2's MARKER0. 
+In extreme cases, underreported will also occur and early conflict detection will not work. 
+Fortunately, we can adjust the frequency of the schedule according to the actual situation to reduce the occurrence of 
+underreported.
+
+Secondly, the marker checker runnable scheduled and selecting the candidate active instants, only the instant smaller than 
+the current writer is selected for conflict detection. This is to avoid two writers discovering conflicts with each 
+other at the same time and failing together.
+
+### [2] Check Commit Conflict: Why we still need to check commit conflict here?
+
+As we know, when a writer completes writing and commits, its corresponding marker information will be deleted 
+immediately. So that, other writers will have underreported in the detection of markers. 
+
+Let's take Figure 4 as an example
+
+![](figure4.png)
+
+Writer1 starts writing data at time t1 and finishes at time t3. Writer2 starts writing at time t2, and will update a 
+file already updated by writer1 at time t4. Since all marker information of writer1 has been deleted at time t4, 
+such conflicts cannot be found in the stage of marker conflict detection until starting to commit. 
+In order to avoid such underreported, it is necessary to add the step of check commit conflict during early conflict 
+detection.
+
+For HoodieMergeHandler we take care about `Latest Base File To Merge` so that we need to make sure this 
+`Latest Base File To Merge` is the same after passed marker conflict checker.
+
+```java
+(table) -> {
+  table.getMetaClient().reloadActiveTimeline();
+  table.getHoodieView().sync();
+  HoodieBaseFile currentBaseFileToMerge = table.getBaseFileOnlyView().getLatestBaseFile(partitionPath, fileId).get();
+  
+  ==> compre currentBaseFileToMerge and oriBaseFileToMerge
+}
+```
+
+For HoodieAppendHandle we take care about `Latest fileSlice` so that we need to make sure this `Latest fileSlice` is 
+the same after passed marker conflict checker.
+
+```java
+(table) -> {
+  table.getMetaClient().reloadActiveTimeline();
+  table.getHoodieView().sync();
+  SliceView currentSliceView = table.getSliceView();
+  Option<FileSlice> currentFileSlice = currentSliceView.getLatestFileSlice(partitionPath, fileId);
+  
+  ==> compre currentFileSlice and oriFileSlice
+}
+```
+
+### [3] Try to resolve conflict
+
+For now, the default behavior is throwing an `HoodieWriteConflictException`
```

Comment:
More details should be added here around how the transaction is ended.  Does the driver or executor throw the exception?  How to gracefully end the current write operation and possibly retry it in the same Spark job with DeltaStreamer?

Reply:
Added